sstables/storage.cc

```diff
@@ -163,20 +198,18 @@ void filesystem_storage::open(sstable& sst) {
 
     // Flushing parent directory to guarantee that temporary TOC file reached
     // the disk.
-    sst.sstable_write_io_check(sync_directory, _dir.native()).get();
+    sst.sstable_write_io_check(sstables::sync_cached_directory, this, _dir.native()).get();
 }
 
 future<> filesystem_storage::seal(const sstable& sst) {
     // SSTable sealing is about renaming temporary TOC file after guaranteeing
     // that each component reached the disk safely.
     co_await remove_temp_dir();
-    auto dir_f = co_await open_checked_directory(sst._write_error_handler, _dir.native());
     // Guarantee that every component of this sstable reached the disk.
-    co_await dir_f.flush();
+    co_await sst.sstable_write_io_check(sstables::sync_cached_directory, this, _dir.native());
```

Comment:
this is a bit cumbersome.
how about passing the error handler to sync_cached_directory?

Reply:
... I tried. :)

Namely, the deeper (preexistent) problem here is that the `filesystem_storage::atomic_delete_complete()` member does not get an `sstable` parameter; therefore, it has no access to the error handler, regardless of how it might want to refer to it (that is, via the `sst._write_error_handler` data member, or the `sst.sstable_write_io_check()` member function). Yet `filesystem_storage::atomic_delete_complete()` still calls `sync_directory()`.

In other words, if you check all the pre-patch call sites of `sync_directory()` and `open_checked_directory()`, in `sstables/storage.cc`, the one in `atomic_delete_complete()` is the only one that does not get access to the error handler. This means that `atomic_delete_complete()` actually cannot notify / signal the rest of the system about `sync_directory()` failing -- regardless of my patch!

Thus, my patch sticks with the "status quo": all call sites except this last one call the friend function (the friend is needed for continued use of `sst.sstable_write_io_check()`), whereas the last call site directly calls the new member function `sync_cached_directory()`.

If the last call site were not an exception wrt. the sstable error handler, I wouldn't  have introduced the friend in the first place. My original approach was precisely what you recommend: pass the error handler into the new member function `sync_cached_directory()` -- what's more, call `open_`**`checked_`**`directory()`, not just `open_directory()`, from within `sync_cached_directory()`. `open_checked_directory()` takes an error handler, and hooks it immediately into the underlying file impl (via `checked_file_impl`). Meaning that I could have eliminated even the preexistent `sst.sstable_write_io_check()` wrapper calls, at all the call sites being converted. However, this breaks down due to `atomic_delete_complete()` not having access to, and being unable to forward, the error handler

Reply:
... I'm not resolving this just yet; I'm curious if this discrepancy between `atomic_delete_complete`, and the other `filesystem_storage` member functions, will actually trigger a parameter list change for `atomic_delete_complete` -- that would have wider-ranging consequences! :)

Reply:
> the one in atomic_delete_complete() is the only one that does not get access to the error handler. This means that atomic_delete_complete() actually cannot notify / signal the rest of the system about sync_directory() failing -- regardless of my patch!

It can pass a "noop" error handler (that could be the default parameter if no error handler is provided explicitly.

BTW, please also take a glimpse into the [brewing future](https://github.com/scylladb/scylladb/pull/19555) of atomic deletion.

> Thus, my patch sticks with the "status quo"

That's acceptable.  I just don't like the passing of `this` in this use case.  It's too C-ish :-/

Reply:
Does `sst.sstable_write_io_check(std::mem_fn(&sstables::sync_cached_directory), _dir.native())` work?
It's the std "textbook" way of passing `[this] (std::string_view name) { return sstables::sync_cached_directory(name); }`

Reply:
> Does `sst.sstable_write_io_check(std::mem_fn(&sstables::sync_cached_directory), _dir.native())` work? It's the std "textbook" way of passing `[this] (std::string_view name) { return sstables::sync_cached_directory(name); }`

It doesn't seem to work. I've tried the following update:

```
diff --git a/sstables/storage.cc b/sstables/storage.cc
index 6480c274dc40..f1dd0a4390ad 100644
--- a/sstables/storage.cc
+++ b/sstables/storage.cc
@@ -202,7 +202,7 @@ void filesystem_storage::open(sstable& sst) {
 
     // Flushing parent directory to guarantee that temporary TOC file reached
     // the disk.
-    sst.sstable_write_io_check(sstables::sync_cached_directory, this, _dir.native()).get();
+    sst.sstable_write_io_check(std::mem_fn(&filesystem_storage::sync_cached_directory), _dir.native()).get();
 }
 
 future<> filesystem_storage::seal(const sstable& sst) {
```

and it produced the following error:

```
In file included from sstables/storage.cc:24:
In file included from ./sstables/sstables_manager.hh:17:
./sstables/sstables.hh:466:16: error: no matching function for call to 'do_io_check'
  466 |         return do_io_check(_write_error_handler, std::forward<Func>(func), std::forward<Args>(args)...);
      |                ^~~~~~~~~~~
sstables/storage.cc:205:9: note: in instantiation of function template specialization 'sstables::sstable::sstable_write_io_check<std::_Mem_fn<seastar::future<> (sstables::filesystem_storage::*)(std::basic_string_view<char>) const>, const std::basic_string<char> &>' requested here
  205 |     sst.sstable_write_io_check(std::mem_fn(&filesystem_storage::sync_cached_directory), _dir.native()).get();
      |         ^
./utils/disk-error-handler.hh:46:1: note: candidate template ignored: substitution failure [with Func = std::_Mem_fn<seastar::future<> (sstables::filesystem_storage::*)(std::basic_string_view<char>) const>, Args = <const std::basic_string<char> &>]: no type named 'type' in 'std::invoke_result<std::_Mem_fn<seastar::future<> (sstables::filesystem_storage::*)(std::basic_string_view<char>) const>, const std::basic_string<char> &>'
   46 | do_io_check(const io_error_handler& error_handler, Func&& func, Args&&... args) {
      | ^
./utils/disk-error-handler.hh:59:6: note: candidate template ignored: constraints not satisfied [with Func = std::_Mem_fn<seastar::future<> (sstables::filesystem_storage::*)(std::basic_string_view<char>) const>, Args = <const std::basic_string<char> &>]
   59 | auto do_io_check(const io_error_handler& error_handler, Func&& func, Args&&... args) noexcept {
      |      ^
./utils/disk-error-handler.hh:57:10: note: because 'std::invocable<std::_Mem_fn<seastar::future<> (sstables::filesystem_storage::*)(std::basic_string_view<char>) const>, const std::basic_string<char> &>' evaluated to false
   57 | requires std::invocable<Func, Args&&...>
      |          ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/14/../../../../include/c++/14/concepts:360:25: note: because 'is_invocable_v<std::_Mem_fn<seastar::future<> (sstables::filesystem_storage::*)(std::basic_string_view<char>) const>, const std::basic_string<char> &>' evaluated to false
  360 |     concept invocable = is_invocable_v<_Fn, _Args...>;
      |                         ^
1 error generated.
```

Reply:
> > the one in atomic_delete_complete() is the only one that does not get access to the error handler. This means that atomic_delete_complete() actually cannot notify / signal the rest of the system about sync_directory() failing -- regardless of my patch!
> 
> It can pass a "noop" error handler (that could be the default parameter if no error handler is provided explicitly.

Assuming the `sync_cached_directory()` member function is supposed to take an error handler, and use a "noop" error handler as default param when it is not passed in (such as when called by `atomic_delete_complete()`):

How can I make the error handler access safe, life-time wise?

`open_checked_directory()` takes a `const io_error_handler&`, and passes it to `make_checked_file()`. The latter also takes a reference, and passes it -- via `make_shared()` argument forwarding -- to the `checked_file_impl()` constructor. Said constructor initializes the `_error_handler`member using a member initializer list, and the type of the `_error_handler` member is *also* a reference.

This means that the `file` object ("checked directory") stored in the map (cache) carries a reference to the error handler object (which is an `std::function` instance). That means those error handlers (both from the `sst._write_error_handler` pass-ins, and the "noop" one) must exist as long as the cache exist. This makes me pretty uncomfortable; I think we should rather create a deep copy of the given error handler, and store that along with the file, in the cache. `std::function` seems light-weight and copying it should not need much CPU or memory. In turn it means we should not use `file` as value type in the map, but a `(file, error handler)` *pair*. What's your take on this? Thanks.

> BTW, please also take a glimpse into the [brewing future](https://github.com/scylladb/scylladb/pull/19555) of atomic deletion.

I apologize; that's just too much for me to take in now. My mental stack is overflowing.